lib/index.js

```diff
@@ -0,0 +1,322 @@
+};
+
+api.parseRequest = (request, options) => {
+  assert.object(request, 'request');
+  assert.object(request.headers, 'request.headers');
+  if(options === undefined) {
+    options = {};
+  }
+  if(options.headers === undefined) {
+    options.headers = [request.headers['x-date'] ? 'x-date' : 'date'];
+  }
+  assert.object(options, 'options');
+  assert.arrayOfString(options.headers, 'options.headers');
+  assert.optionalFinite(options.clockSkew, 'options.clockSkew');
+
+  const authzHeaderName = options.authorizationHeaderName || 'authorization';
+
+  if(!request.headers[authzHeaderName]) {
+    throw new HttpSignatureError('no ' + authzHeaderName + ' header ' +
+      'present in the request', 'DataError');
+  }
+
+  options.clockSkew = options.clockSkew || 300;
+
+  const ParamsState = {
+    Name: 0,
+    Quote: 1,
+    Value: 2,
+    Comma: 3
+  };
+
+  const State = {
+    New: 0,
+    Params: 1
+  };
+
+  let state = State.New;
+  let substate = ParamsState.Name;
+  let tmpName = '';
+  let tmpValue = '';
+
+  const parsed = {
+    scheme: '',
+    params: {},
+    signingString: ''
+  };
+
+  const authz = request.headers[authzHeaderName];
+  for(let i = 0; i < authz.length; i++) {
+    const c = authz.charAt(i);
+
+    switch(Number(state)) {
+
+      case State.New:
+        if(c !== ' ') {
+          parsed.scheme += c;
+        } else {
+          state = State.Params;
+        }
+        break;
+
+      case State.Params:
+        switch(Number(substate)) {
+
+          case ParamsState.Name:
+            const code = c.charCodeAt(0);
+            // restricted name of A-Z / a-z
+            if((code >= 0x41 && code <= 0x5a) || // A-Z
+                (code >= 0x61 && code <= 0x7a)) { // a-z
+              tmpName += c;
+            } else if(c === '=') {
+              if(tmpName.length === 0) {
+                throw new HttpSignatureError('bad param format', 'DataError');
+              }
+              substate = ParamsState.Quote;
+            } else {
+              throw new HttpSignatureError('bad param format', 'DataError');
+            }
+            break;
+
+          case ParamsState.Quote:
+            if(c === '"') {
+              tmpValue = '';
+              substate = ParamsState.Value;
+            } else {
+              throw new HttpSignatureError('bad param format', 'DataError');
+            }
+            break;
+
+          case ParamsState.Value:
+            if(c === '"') {
+              parsed.params[tmpName] = tmpValue;
+              substate = ParamsState.Comma;
+            } else {
+              tmpValue += c;
+            }
+            break;
+
+          case ParamsState.Comma:
+            if(c === ',') {
+              tmpName = '';
+              substate = ParamsState.Name;
+            } else {
+              throw new HttpSignatureError('bad param format', 'DataError');
+            }
+            break;
+
+          default:
+            throw new Error('Invalid substate');
```

Comment:
Should these by `HTTPSignatureError`s? Specifically of the type `InvalidState`? Why are they only `Error`?